src/core/net/dnssd_server.cpp

```diff
@@ -1103,8 +1140,39 @@ void Server::FinalizeQuery(QueryTransaction &aQuery, Header::Response aResponseC
     OT_ASSERT(sdType != kDnsQueryNone);
     OT_UNUSED_VARIABLE(sdType);
 
-    mQueryUnsubscribe(mQueryCallbackContext, name);
     aQuery.Finalize(aResponseCode, mSocket);
+    // We are the last subscribing this name
+    if (!HasSameQuery(name, sdType))
+    {
+        mQueryUnsubscribe(mQueryCallbackContext, name);
+    }
+}
+
+bool Server::HasSameQuery(const char *name, DnsQueryType aQueryType)
+{
+    const char *split = GetDomainSplit(name, aQueryType);
+    bool        found = false;
+
+    for (QueryTransaction &query : mQueryTransactions)
+    {
+        char         queryName[Name::kMaxNameSize];
```

Comment:
```suggestion
```
I think it's possible to implement query transaction searching without using another buffer. 
`Dns::Name::CompareName` allows comparing C strings with DNS names in messages without reading to buffers. 

BTW, We have been very careful on allocating buffers of `char [Name::kMaxNameSize]` in the callchains of DNS-SD server to avoid increasing stack usage.